### ISSUE TITLE
Add support for CAMT / ISO 20022 import

### ISSUE THREAD

**Issue by lorenzschmid** [firefly-iii/firefly-iii#5098]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)
- [X] My request is not listed as [a very good idea, but unfortunately...](https://docs.firefly-iii.org/firefly-iii/about-firefly-iii/what-its-not/)
- [X] I've used [the search](https://github.com/firefly-iii/firefly-iii/issues?q=is%3Aissue) and this has not been discussed before.

### Description

I would like to have an importer which allows to import CAMT files (ISO 200222 standard) commonly used by banks:
- Bank-to-Customer Account Report (camt.052)  
- Bank-to-Customer Statement (camt.053) 
- Bank-to-Customer Debit/Credit Notification (camt.054)

More details on the format can be found [here](https://www.six-group.com/dam/download/banking-services/interbank-clearing/en/standardization/iso/swiss-recommendations/archives/implementation-guidelines-cm/implementation-guidelines-camt_v1_6_181217.pdf) (although written for the Swiss banking exchange, section 3 to 5 explain the three formats above). There was a previous issue #1932 on this subject which was closed without any implementation.

### Solution

Add a new/extend an existing importer to support CMAT files.

**Comment by JC5** [comment]
I'll put it on the list!

**Comment by cstuder** [comment]
There seems to be an existing PHP library able to parse these XML files: https://github.com/genkgo/camt.

My bank (PostFinance) is about to shut down their analytics tool, Firefly III would be a perfect replacement and this is one of the export formats they offer.

**Comment by cstuder** [comment]
@JC5 If I wanted to extend Firefly III to be able to import camt.053-files, could you point me in the right direction for development? Do you have an overview how services in the [data-importer](https://github.com/firefly-iii/data-importer/tree/main/app/Services) work? What kind of interface do they need to support?

**Comment by JC5** [comment]
For a start, check out routes/web.php and see the steps the importer goes through. Basically, for the CSV import, it goes like this:

- step 3 (yes), upload file, see app/Http/Controllers/UploadController
- step 4, ConfigurationController
- step 5, RoleController. This step may not be necessary for camt.053, depending on their format.
- step 6, MapController

Step 7: The ConversionController takes the config + uploaded file and converts it into Firefly III compatible arrays. You'll see that for each flow a different RoutineManager is created. After this point, the submission routine is the same for all flows because each flow is expected to create Firefly III compatible arrays.

You can clone the CSV routine manager and stick with that. 

The CSV routine is pretty complex because it'll convert each individual cell according to its column's role, and then process any mapping, look-ups, etc.

I could make the 'csv' flow a generic 'file' flow and give it some params, but that presumes that all steps are the same.

**Comment by beatbesmer** [comment]
Anyone interessted in this: can you provide a list of fields from camt.053 (please correct Name according specificaction of ISO20022), which should be extracted and provide to the importer.

**Comment by cstuder** [comment]
Still interested, but haven't found the time yet.

I've started with some data exploration by writing a CSV converter, you can find it here: https://github.com/cstuder/camt2csv (Work in progress, comment line 71 to make it work in the current state)

The repo includes a real world (anonymised & randomised) camt file, generated by a PostFinance personal account.

I've tried matching relevant camt fields to Firefly columns, but the matching is not yet perfect. Some of the fields are empty. I assume that depending on the transaction type you would have to take different camt fields and do some guesswork.

**Comment by beatbesmer** [comment]
After some back and forth, I think I will let the users assign the roles themselves. In between, I wanted to skip that part completely and hard code the role assignment. With the current workflow, we create an array with many camt.053 fields (too many). But by going this route, users have full flexibility as with a CSV file, but can still export their preferred configuration.xml for future use to save assigning 1001 roles ;-)

**Comment by plumped** [comment]
Der All,

I have built a Python script to parse the offical CAMT / ISO 20022 Reports from Europe Banks. Please find it attached and maybe  you can use it in firefly importer itself.

1. Set the folder path (all *.xml files will be grabbed and parsed at once in the given folder (Multiple account reports for example)
2. Check and replace the Regex Expression for your specific Country IBAN (Switzerland allready given)
3. Run script
4. Import formated csv file in FIDI

Enjoy!

```
import csv
import glob
import json
import re
from datetime import datetime

from sepa import parser

# Utility function to remove additional namespaces from the XML
def strip_namespace(xml):
    return re.sub(' xmlns="[^"]+"', '', xml, count=1)

# Loop over all XML files in the directory
for file_path in glob.glob('*.xml'):
    # Read file
    with open(file_path, 'r') as f:
        input_data = f.read()

    # Parse the bank statement XML to JSON
    camt53 = parser.parse_string(parser.bank_to_customer_statement, bytes(strip_namespace(input_data), 'utf8'))

    # Pretty print JSON information from bank statement
    #print(json.dumps(camt53, indent=4))
    #print()

    with open(file_path.replace('.xml', '.csv'), 'w') as csv_file:
        csv_writer = csv.writer(csv_file)

        csv_writer.writerow(['Account', 'Creditor Name', 'Creditor IBAN', 'Reference', 'Transaction Details', 'Credit/debit', 'Amount', 'Currency', 'Status', 'Booking Date', 'Value Date'])
        
        
        #Find Function    
        def find_in_string(result, regex):
            search_for_string = re.search(regex, result)
            if (search_for_string is None):
                search_for_string = ""
                return search_for_string
            else:
                return search_for_string.group()
        
        
        # Process bank statements
        for statement in camt53['statements']:
            if 'entries' not in statement:
                continue

            for entry in statement['entries']:
                csv_writer.writerow([
                    statement['account']['id']['iban'],
                
                #Get Creditors Name    
                    find_in_string(str(entry['entry_details']), "(?<='creditor': {'name': ')(.*)(?=')"),
                
                #Get destination IBAN-Number
                    find_in_string(str(entry['entry_details']), "CH\d{2}[ ]\d{4}[ ]\d{4}[ ]\d{4}[ ]\d{4}[ ]\d{1}|CH\d{19}"),
                    entry['account_servicer_reference'],
                
                #Get Remittance Information
                    find_in_string(str(entry['entry_details']), "(?<='remittance_information': {'unstructed': \[\')(.*)(?='])"),
                
                #Get Credit/Debit indicator
                    entry['credit_debit_indicator'],
                
                #Get Amount
                    entry['amount']['_value'],
                    
                #Get Currency
                    entry['amount']['currency'],
                    
                #Get Booking status
                    entry['status'],
                    
                #Get Booking date    
                    entry['booking_date']['date'],
                    
                #Get Value Date
                    entry['value_date']['date'],
                    
                ]
                )

```

**Comment by JC5** [comment]
You should post this on [Gist](https://gist.github.com/), and link to it 👍🏻 . That way, people can comment and you can easily edit it!

**Comment by plumped** [comment]
https://gist.github.com/plumped/61bd50314414ae760991eb233f10d105

**Comment by plumped** [comment]
@JC5 You already have the whole thing in progress according to your project plan, when do you think the banking standard will be available? Are you familiar with the subject matter? As a Dutchman you would also benefit from it :D

**Comment by JC5** [comment]
There is a lot on my roadmap, so unfortunately I can't give out estimations :(

**Comment by plumped** [comment]
Hey @JC5 
I have tried some different finance tools where CAMT import is already implemented. But at the end i always come back to firefly because of the look, the feeling and your effort within this project. Absolutely brilliant! 

But the csv import **still** sucks :D every financial institution in europe now uses ISO20022 with the camt standards, because there are simply more options to provide more data within the file type. it would be simply THE SHIT to import it directly finally.

I also massage your balls for easter if needed xD

HAPPY EASTER!

**Comment by beatbesmer** [comment]
@plumped nice offer :) Genkgo camt is used in my attempt already. Unfortunately, there is not much time to knitt something useful and simple to use. There isn't a stable version online, but if someone wants to join / help, I can try to clean the current working directory so we could use this branch to continue work in a team. My 'one man show' with family seems a bad idea

**Comment by plumped** [comment]
Hi @beatbesmer 
I don't think we need to reinvent the wheel. The Genkgo attempt is already working perfectly with camt.53 and .54 as well which are the „standard“ for customer communications. The only thing that would need to be done would be the integration into the importer, and the mapping of the outputfields.
1. Problem: im not a PHP dev
2. Problem: im not even a dev and just coding as a hobbie between job and family.
3. The knowledge for a clean integration into his application has only @JC5 

Of course we could use the API for db entries but at the end its just another workaround :)

**Comment by JC5** [comment]
> 3. The knowledge for a clean integration into his application has only @JC5

It's not *that* difficult. I believe the code for the importer isn't that hard to follow.

**Comment by beatbesmer** [comment]
@plumped 
Well, if it were that simple, this would have been implemented a long time ago. :)
Sure, FireFly offers many possibilities to do this easily, but you still have to learn the ropes. Another difficulty seems to be (as James already mentioned in his news once) that the banks do not necessarily implement the standard uniformly and still have some leeway. And thus, of course, each import must be adaptable by each user, his bank and his wishes. This costs additional time to implement these possibilities.

I'm trying my hand at [a fork](https://github.com/beatbesmer/data-importer/tree/develop) and will commit and push each time I have partial success, so that this can theoretically be tested by everyone. Add yourself as a Watcher in the following to stay up to date.
As I am sometimes struggling with GIT, there may be a bit more messages sent than necessary...
Feel free to add Ideas / FeatureRequest etc. as Issues

**Comment by beatbesmer** [comment]
> next i will set up a zip reader so that not every file has to be read individually (let you know when its done for usage). when this is up and running i will interface or create a web app with access to the Firefly-iii API.
> 
> Have a look at the code if you are interested: https://github.com/plumped/camt_converter_ISO20022_for_camt

@plumped 
Without looking at your code, multifile upload and zip extraction / reading was a topic of a discussion with @JC5 last year. Not sure if some code made it into the branch 'multi-file'. 

Maybe some code from my lost experiments help. This was used for filetype detection, should be changeable to read files from a zip

```php
<?php
$file = $argv[1];
$fileType = mime_content_type($file);
echo "Filetype of File '".$file."' is '".$fileType."'\n";
if($fileType == 'application/zip') {
     $archive = new ZipArchive();
     $archive->open($file);
     for($index = 0; $index != $archive->count(); $index++) {
         $singleFileName = $archive->getNameIndex($index);
         echo "Filename is: ".$singleFileName."\n";
         //$singleFileStream = $archive->getStream($index);
         $singleFileContent = $archive->getFromIndex($index);
         echo file_put_contents(
             $singleFileName,
             $singleFileContent,
             $flags = 0,
             //$context = $singleFileStream
         );
         $singleFileType = mime_content_type($singleFileName);
         echo "FileType is: ".$singleFileType."\n";
     }
}
```

I'll check your code ASAP. Thx in advance

**Comment by plumped** [comment]
@beatbesmer Good morning Beat

thanks for your code! The possibility to load zips is already implemented since last commit.
1. load zip with camt
2. Same procedure is starting and gives a list with all ntrys from all files as output.

Next i just need to export it as csv for a workarround until the interface is coded

**Comment by plumped** [comment]
Dear @JC5 @lorenzschmid @cstuder 

I have created a fully functional CAMT Converter which takes a *.zip file and converts to collected iban.csv as a docker webapplication.

https://github.com/plumped/camt_converter_ISO20022_for_camt 
As of today, it is just supported for Swiss IBANS beginning with CH#, but will add multifunctionality this week, or tonight :D.
As a future target i'll try to feed it directly to firefly over the API interface.

**Comment by JC5** [comment]
Nice work! Shall I add it to the documentation?

**Comment by plumped** [comment]
> Nice work! Shall I add it to the documentation?

Of course, feel free!

**Comment by bakeromso** [comment]
I have a bank export in camt053 from Triodos. The XML file contains `<Document xmlns="urn:iso:std:iso:2002
2:tech:xsd:camt.053.001.02">`, so I suppose it should follow the ISO 20022 standard. I inspected this file manually and used @plumped's script to be sure, but the export from the bank is still as unclean as the CSV file. When taking money from an ATM for example, the date of transaction is not when money was drawn, but when the bank processed the transaction. The opposing IBAN is empty, but placed in the description. Is this a problem caused by the bank's export? 

If so, I am working on a Python script to clean that up which, I think, can be valuable to others. What would be a sensible place to share this?

**Comment by JC5** [comment]
> but the export from the bank is still as unclean as the CSV file

Usually, when banks have shitty exports in one format, the other formats are also terrible.

**Comment by bakeromso** [comment]
> > but the export from the bank is still as unclean as the CSV file
> 
> Usually, when banks have shitty exports in one format, the other formats are also terrible.

Then I shall proceed with my cleanup script! Do you have any ideas how I can make it so that other firefly users can use it easily as well? I could make something like the CAMT importer from @plumped, that users can add to their docker file, cleaning their CSV? Or can you think of some other way that maybe makes it more streamlined?

**Comment by plumped** [comment]
@JC5 can you please give me an advise how a pattern has to look like for the transaction creation on api. I already read the swagger documentation, but im not sure what exactly i have to transfer for a creation. Now that my code is optimized and ready to run, I'll take on the API interface. Thanks in advance for a template

**Comment by JC5** [comment]
Could you open a separate discussion for that please?

**Comment by JC5** [comment]
And it's in place!
